### PR TITLE
.Net: MongoDB memory store connector for .NET

### PR DESCRIPTION
**Pull request by BorisDog**
### Motivation and Context
Resolves: #2817

### Description

MongoDB memory store connector for .NET, build on top of MongoDB .NET driver. A memory store abstraction that facilitates connections between a MongoDB Atlas cluster to conduct [Atlas Vector Search](https://www.mongodb.com/products/platform/atlas-vector-search) on the Microsoft Semantic Kernel.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by BorisDog** [comment]
@microsoft-github-policy-service agree company="MongoDB"

File: dotnet/src/Connectors/Connectors.Memory.MongoDB/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+﻿# Microsoft.SemanticKernel.Connectors.Memory.MongoDB
```
Comment (Jibola) [inline dotnet/src/Connectors/Connectors.Memory.MongoDB/README.md:1]: I think we should also add in the README how to configure the cluster to pass our tests as well. That may be useful for folks.

**Comment by kbeaugrand** [comment]
@BorisDog, I'm a little sad to see you doing what I did a while back ;(


[https://github.com/kbeaugrand/SemanticKernel.Connectors.Memory.MongoDB](https://github.com/kbeaugrand/SemanticKernel.Connectors.Memory.MongoDB)

Honestly, that's what I meant when I talked to @alexchaomander about integrating community connectors into SK directly. As I said, I'd rather create my connector and create a PR for them directly instead of publishing it myself. That way, more people can benefit from it, rather than me doing work that was useful for me but won't be reused by others (or, in this case, even redone).

**Comment by prakul** [comment]
@shawncal can you please review the PR whenever possible.

**Comment by markwallace-microsoft** [comment]
@BorisDog one typo to be fixed

```
Warning: "Intergation" should be "Integration".
Warning: "Intergation" should be "Integration".
error: `Intergation` should be `Integration`
  --> ./dotnet/src/IntegrationTests/Connectors/Memory/MongoDB/MongoDBMemoryStoreTestsFixture.cs:45:[43](https://github.com/microsoft/semantic-kernel/actions/runs/6735614169/job/18309170062?pr=2950#step:3:45)
   |
[45](https://github.com/microsoft/semantic-kernel/actions/runs/6735614169/job/18309170062?pr=2950#step:3:47) |         this.DatabaseTestName = "dotnetMSKIntergationTests1";
   |                                           ^^^^^^^^^^^
   |
error: `Intergation` should be `Integration`
  --> ./dotnet/src/IntegrationTests/Connectors/Memory/MongoDB/MongoDBMemoryStoreTestsFixture.cs:[46](https://github.com/microsoft/semantic-kernel/actions/runs/6735614169/job/18309170062?pr=2950#step:3:48):58
   |
46 |         this.ListCollectionsDatabaseTestName = "dotnetMSKIntergationTests2";
   |                                                          ^^^^^^^^^^^
```

**Comment by BorisDog** [comment]
Thanks @markwallace-microsoft, done.

File: dotnet/src/IntegrationTests/testsettings.json [hunk 2]
```diff
@@ -30,5 +30,9 @@
   },
   "Postgres": {
     "ConnectionString": ""
+  },
+  "MongoDB": {
+    "ConnectionString": "",
+    "VectorSearchCollection": "dotnetMSKNearestTest.nearestSearch"
```
Comment (markwallace-microsoft) [inline dotnet/src/IntegrationTests/testsettings.json:36]: @BorisDog  How do I set up this collection for testing?

**Review by markwallace-microsoft — COMMENTED** [review]
The code LGTM

I followed the steps to set up the connector and tried to run the integration tests but this failed

image

File: dotnet/src/IntegrationTests/testsettings.json (hunk 2 above)
Comment (BorisDog) [reply dotnet/src/IntegrationTests/testsettings.json:36]: You would need to setup the following index via Atlas UI:
```
{
  "mappings": {
    "dynamic": true,
    "fields": {
      "embedding": {
        "dimensions": 3,
        "similarity": "cosine",
        "type": "knnVector"
      }
    }
  }
}
```
The collection is populated in `MongoDBMemoryStoreTestsFixture.InitializeAsync()`
Or alternatively you could use the Atlas cluster that was set up for these integration tests.

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/samples/KernelSyntaxExamples/Example15_TextMemoryPlugin.cs [hunk 3]
```diff
@@ -84,6 +88,12 @@ private static async Task<IMemoryStore> CreateSampleDuckDbMemoryStoreAsync()
         return store;
     }
 
+    private static IMemoryStore CreateSampleMongoDBMemoryStore()
+    {
+        IMemoryStore store = new MongoDBMemoryStore(TestConfiguration.MongoDB.ConnectionString, "memorySkillExample");
```
Comment (dmytrostruk) [inline dotnet/samples/KernelSyntaxExamples/Example15_TextMemoryPlugin.cs:93]: nit: 
```suggestion
        IMemoryStore store = new MongoDBMemoryStore(TestConfiguration.MongoDB.ConnectionString, "memoryPluginExample");
```

File: dotnet/src/Connectors/Connectors.Memory.MongoDB/MongoDBMemoryEntry.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    /// Returns mapped <see cref="MemoryRecord"/>.
+    /// </summary>
+    public MemoryRecord ToMemoryRecord() =>
+        new(this.Metadata.ToMemoryRecordMetadata(), this.Embedding, this.Id, this.Timestamp?.ToLocalTime());
+
+    /// <summary>
+    /// Returns a pair of mapped <see cref="MemoryRecord"/> and score.
+    /// </summary>
+    public (MemoryRecord, double) ToMemoryRecordAndScore() =>
+        (new(this.Metadata.ToMemoryRecordMetadata(), this.Embedding, this.Id, this.Timestamp?.ToLocalTime()), this.Score);
```
[the remark this answers is not in the document]
Comment (markwallace-microsoft) [reply dotnet/src/Connectors/Connectors.Memory.MongoDB/MongoDBMemoryEntry.cs:72]: @BorisDog could you take a look at the comments. Thanks.

File: dotnet/src/Connectors/Connectors.Memory.MongoDB/MongoDBMemoryStore.cs [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="MongoDBMemoryStore" /> class.
+    /// </summary>
+    /// <param name="mongoClient">MongoDB client.</param>
+    /// <param name="databaseName">Database name.</param>
+    /// <param name="indexName">Name of the search index. If no value is provided default index will be used.</param>
+    public MongoDBMemoryStore(IMongoClient mongoClient, string databaseName, string? indexName = default)
```
[the remark this answers is not in the document]
Comment (markwallace-microsoft) [reply dotnet/src/Connectors/Connectors.Memory.MongoDB/MongoDBMemoryStore.cs:35]: https://github.com/microsoft/semantic-kernel/issues/3431

**Pull request merged**